Create export_for_inference API and expose core_aten as public facing API

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #135918
* __->__ #135912
* #135080

Differential Revision: [D62606908](https://our.internmc.facebook.com/intern/diff/D62606908)

tugsbayasgalan committed:
Create export_for_inference API

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D62606908).

avikchaudhuri commented on torch/export/__init__.py:
```diff
@@ -165,6 +173,91 @@ def export_for_training(
+    kwargs: Optional[Dict[str, Any]] = None,
+    *,
+    dynamic_shapes: Optional[Union[Dict[str, Any], Tuple[Any], List[Any]]] = None,
+    decomp_table: Optional[Dict["OpOverload", Optional[Callable]]] = None,
```
Perhaps this should go last?

avikchaudhuri commented on torch/export/exported_program.py:
```diff
@@ -283,6 +284,17 @@ def _split_decomp_table_to_cia_and_python_decomp(
 
+def core_aten_decompositions() -> Dict[torch._ops.OperatorBase, Callable]:
+    """
+    This is an official decomposition table which contains decomposition of
```
Cut `official`? You can say "default." Or:

This returns a table of decompositions for all aten operators to core aten operators. Use this API...

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D62606908).

tugsbayasgalan committed:
Update on "Create export_for_inference API and expose core_aten as public facing API"


Differential Revision: [D62606908](https://our.internmc.facebook.com/intern/diff/D62606908)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D62606908).

tugsbayasgalan committed:
Update on "Create export_for_inference API and expose core_aten as public facing API"


Differential Revision: [D62606908](https://our.internmc.facebook.com/intern/diff/D62606908)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D62606908).

tugsbayasgalan committed:
Update on "Create export_for_inference API and expose core_aten as public facing API"


Differential Revision: [D62606908](https://our.internmc.facebook.com/intern/diff/D62606908)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D62606908).

tugsbayasgalan committed:
Update on "Create export_for_inference API and expose core_aten as public facing API"


Differential Revision: [D62606908](https://our.internmc.facebook.com/intern/diff/D62606908)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Create export_for_inference API and expose core_aten as public facing API"


Differential Revision: [D62606908](https://our.internmc.facebook.com/intern/diff/D62606908)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D62606908).

tugsbayasgalan committed:
Update on "Create export_for_inference API and expose core_aten as public facing API"


Differential Revision: [D62606908](https://our.internmc.facebook.com/intern/diff/D62606908)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Create export_for_inference API and expose core_aten as public facing API"


Differential Revision: [D62606908](https://our.internmc.facebook.com/intern/diff/D62606908)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D62606908).

tugsbayasgalan committed:
Update on "Create export_for_inference API and expose core_aten as public facing API"


Differential Revision: [D62606908](https://our.internmc.facebook.com/intern/diff/D62606908)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D62606908).

tugsbayasgalan committed:
Update on "Create export_for_inference API and expose core_aten as public facing API"


Differential Revision: [D62606908](https://our.internmc.facebook.com/intern/diff/D62606908)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D62606908).

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D62606908).

tugsbayasgalan committed:
Update on "Create export_for_inference API and expose core_aten as public facing API"


Differential Revision: [D62606908](https://our.internmc.facebook.com/intern/diff/D62606908)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Create export_for_inference API and expose core_aten as public facing API"


Differential Revision: [D62606908](https://our.internmc.facebook.com/intern/diff/D62606908)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D62606908).
